src/Features/CSharp/Portable/CodeRefactorings/NamedArguments/CSharpAddNamedArgumentsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CodeRefactorings.NamedArguments;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeRefactorings.NamedArguments
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpAddNamedArgumentsCodeRefactoringProvider)), Shared]
+    internal class CSharpAddNamedArgumentsCodeRefactoringProvider : AbstractAddNamedArgumentsCodeRefactoringProvider<ArgumentSyntax>
+    {
+        protected override IEnumerable<ArgumentSyntax> GetNamedArguments(IList<ArgumentSyntax> arguments, ImmutableArray<IParameterSymbol> parameters)
+        {
+            if (arguments == null)
+            {
+                throw new ArgumentNullException(nameof(arguments));
+            }
+
+            for (int index = 0; index < arguments.Count; ++index)
+            {
+                var parameter = parameters[index];
+                var argument = arguments[index];
+
+                if (argument.NameColon != null)
+                {
+                    yield return argument;
+                }
+                else if (parameter.IsParams && arguments.Count - index > 1)
+                {
+                    var expressions = arguments.Skip(index).Select(arg => arg.Expression);
+                    var arrayCreation = WrapInArrayCreation(expressions);
+                    var namedArgument = Argument(arrayCreation).WithNameColon(NameColon(parameter.Name));
+                    yield return namedArgument;
+                    yield break;
+                }
+                else
+                {
+                    yield return argument.WithNameColon(NameColon(parameter.Name));
+                }
+            }
+        }
+
+        private static ImplicitArrayCreationExpressionSyntax WrapInArrayCreation(IEnumerable<ExpressionSyntax> expressions)
+        {
+            var initializer = InitializerExpression(SyntaxKind.ArrayInitializerExpression, SeparatedList(expressions));
+            return ImplicitArrayCreationExpression(initializer);
```

Comment:
I think this may end up failing.  For example, if you had:

```
void Foo(params IComparable[] args) { ... }
Foo(new Type1(), new Type2());
```

If you replace that with ```Foo(new [] { new Type1(), new Type2() })``` you can end up with the compiler saying "there is no best common type between 'Type1' and 'Type2'"